### PR TITLE
Add SArray.hash() and SArray.random_integers()

### PR DESCRIPTION
**Pull request by ylow**
These are very useful for things like exact subsampling, sorting, etc.

### CODE REVIEW

File: oss_src/unity/lib/unity_sarray.cpp [hunk 1]
```diff
@@ -1622,6 +1622,16 @@ std::shared_ptr<unity_sarray_base> unity_sarray::sample(float percent,
   return logical_filter(seq);
 }
 
+
+std::shared_ptr<unity_sarray_base> unity_sarray::hash(int random_seed) {
+  flex_int seed_hash = flexible_type((flex_int)(random_seed)).hash();
+  auto filter_fn = [seed_hash](const flexible_type& val)->flexible_type {
+        return hash64(val.hash() ^ seed_hash);
```
Comment (hoytak) [inline oss_src/unity/lib/unity_sarray.cpp]: Do hash64(val.hash(), seed_hash); it's set up for this.

File: oss_src/unity/lib/unity_sarray.cpp [hunk 2]
```diff
@@ -1622,6 +1622,16 @@ std::shared_ptr<unity_sarray_base> unity_sarray::sample(float percent,
   return logical_filter(seq);
 }
 
+
+std::shared_ptr<unity_sarray_base> unity_sarray::hash(int random_seed) {
+  flex_int seed_hash = flexible_type((flex_int)(random_seed)).hash();
```
Comment (hoytak) [inline oss_src/unity/lib/unity_sarray.cpp]: just do hash64(random_seed)

File: oss_src/unity/lib/unity_sarray.cpp [hunk 3]
```diff
@@ -1622,6 +1622,16 @@ std::shared_ptr<unity_sarray_base> unity_sarray::sample(float percent,
   return logical_filter(seq);
 }
 
+
+std::shared_ptr<unity_sarray_base> unity_sarray::hash(int random_seed) {
```
Comment (hoytak) [inline oss_src/unity/lib/unity_sarray.cpp]: why int?  Shouldn't this be size_t?

File: oss_src/unity/lib/unity_sarray.hpp [hunk 4]
```diff
@@ -515,6 +515,12 @@ class unity_sarray: public unity_sarray_base {
    */
   std::shared_ptr<unity_sarray_base> sample(float percent, int random_seed);
 
+  /**
+   * Returns an SArray of type flex_int that contains the hash of each element.
+   * The hash function takes a seed value so this can be used for
+   * random generation as well.
+   */
+  std::shared_ptr<unity_sarray_base> hash(int seed);
```
Comment (hoytak) [inline oss_src/unity/lib/unity_sarray.hpp]: ditto - int vs. size_t.

File: oss_src/unity/python/sframe/cython/cy_sarray.pxd [hunk 5]
```diff
@@ -64,6 +64,7 @@ cdef extern from "<unity/lib/api/unity_sarray_interface.hpp>" namespace 'graphla
         unity_sarray_base_ptr drop_missing_values() except +
         unity_sarray_base_ptr fill_missing_values(flexible_type) except +
         unity_sarray_base_ptr sample(float, int) except +
+        unity_sarray_base_ptr hash(int) except +
```
Comment (hoytak) [inline oss_src/unity/python/sframe/cython/cy_sarray.pxd]: ditto

File: oss_src/unity/python/sframe/cython/cy_sarray.pxd [hunk 6]
```diff
@@ -175,6 +176,8 @@ cdef class UnitySArrayProxy:
 
     cpdef sample(self, float percent, int seed)
 
+    cpdef hash(self, int seed)
```
Comment (hoytak) [inline oss_src/unity/python/sframe/cython/cy_sarray.pxd]: ditto. or just do a python object, then call hash() on that.  then it's a 64bit int.

File: oss_src/unity/python/sframe/data_structures/sarray.py [hunk 7]
```diff
@@ -1873,6 +1873,36 @@ def sample(self, fraction, seed=None):
...
+        with cython_context():
+            return SArray(_proxy=self.__proxy__.hash(seed))
+
+    @classmethod
+    def random_integers(self, size, seed=None):
+        """
+        Returns an SArray with random integer values.
+        """
+        if seed is None:
+            seed = time.time()
```
Comment (hoytak) [inline oss_src/unity/python/sframe/data_structures/sarray.py]: This has resolution up to a second, so calls within the same second produce the same values.  Use ` seed = hash("%0.20f" % time.time())`.

File: oss_src/unity/python/sframe/test/test_sarray.py [hunk 8]
```diff
@@ -977,6 +977,31 @@ def test_sample(self):
         sa_sample = SArray().sample(.5, 9)
         self.assertEqual(len(sa_sample), 0)
 
+    def test_hash(self):
```
Comment (hoytak) [inline oss_src/unity/python/sframe/test/test_sarray.py]: Add test that two subsequent calls with seed = None do not have the same values.

Comment (ylow) [inline oss_src/unity/python/sframe/test/test_sarray.py]: hash is fully deterministic. seed=None does not make sense for a hash method.

File: oss_src/unity/lib/unity_sarray.cpp (hunk 3 above)
Comment (ylow) [inline oss_src/unity/lib/unity_sarray.cpp]: Was going to. But we have "int" *everywhere* going quite deep. sample takes an int as well.
Fixing that will be a much larger code change than I want to do now.

File: oss_src/unity/lib/unity_sarray.hpp (hunk 4 above)
Comment (ylow) [inline oss_src/unity/lib/unity_sarray.hpp]: Can't change to size_t now. int goes quite deep.

File: oss_src/unity/lib/unity_sarray.cpp (hunk 3 above)
Comment (hoytak) [inline oss_src/unity/lib/unity_sarray.cpp]: Ah, okay.  Let's maybe then take it and do `seed = hash(seed) % 2**31` at the python level.

File: oss_src/unity/python/sframe/data_structures/sarray.py (hunk 7 above)
Comment (ylow) [inline oss_src/unity/python/sframe/data_structures/sarray.py]: At the moment, this is consistent with how sa.sample() works.I could fix it there. But then it will definitely be necessary to expand int to size_t. (or ssize_t). which will be a large change.

File: oss_src/unity/lib/unity_sarray.cpp (hunk 3 above)
Comment (hoytak) [inline oss_src/unity/lib/unity_sarray.cpp]: Reference: https://github.com/dato-code/SFrame/blob/master/oss_src/unity/python/sframe/data_structures/sframe.py#L3241

File: oss_src/unity/python/sframe/data_structures/sarray.py (hunk 7 above)
Comment (hoytak) [inline oss_src/unity/python/sframe/data_structures/sarray.py]: Ref: https://github.com/dato-code/SFrame/blob/master/oss_src/unity/python/sframe/data_structures/sframe.py#L3241

File: oss_src/unity/python/sframe/test/test_sarray.py (hunk 8 above)
Comment (hoytak) [inline oss_src/unity/python/sframe/test/test_sarray.py]: Ah, sorry, I meant on random_integers.

File: oss_src/unity/lib/unity_sarray.cpp (hunk 3 above)
Comment (ylow) [inline oss_src/unity/lib/unity_sarray.cpp]: Oddly: https://github.com/dato-code/SFrame/blob/master/oss_src/unity/python/sframe/data_structures/sframe.py#L3192

Comment (ylow) [inline oss_src/unity/lib/unity_sarray.cpp]: Then sample should probably be fixed too.

File: oss_src/unity/python/sframe/data_structures/sarray.py [hunk 9]
```diff
@@ -1952,12 +1952,42 @@ def sample(self, fraction, seed=None):
...
+        -------
+        out : SArray
+            An integer SArray with a hash value for each element. Identical
+            elements are hashed to the same value
+        """
+        with cython_context():
+            return SArray(_proxy=self.__proxy__.hash(seed))
+
+    @classmethod
+    def random_integers(self, size, seed=None):
```
Comment (TobyRoseman) [inline oss_src/unity/python/sframe/data_structures/sarray.py]: When it's a classmethod, the first parameter shouldn't be `self`, usually `cls` is used.

Comment (TobyRoseman) [inline oss_src/unity/python/sframe/data_structures/sarray.py]: Maybe `length` rather than `size`.

File: oss_src/unity/python/sframe/data_structures/sarray.py [hunk 10]
```diff
@@ -1952,12 +1952,42 @@ def sample(self, fraction, seed=None):
...
+        """
+        Returns an SArray with random integer values.
+        """
+        if seed is None:
+            seed = abs(hash("%0.20f" % time.time())) % (2 ** 31)
+        return SArray.from_sequence(size).hash(seed)
```
Comment (TobyRoseman) [inline oss_src/unity/python/sframe/data_structures/sarray.py]: Assuming you take my previous comment you probably want to change "SArray" to "cls". At any rate I don't think you should have "SArray" hard coded.

**Comment by ylow** [comment]
@TobyRoseman Yup. done

**Comment by TobyRoseman** [comment]
Looks good to me
